### ISSUE TITLE
关于Priors和数据增广

### ISSUE THREAD

**Issue by dongfangduoshou123** [aoru45/LFFD-Pytorch#1]
感谢基于ｐｙｔｏｒｃｈ的复现！
我有两个问题想请教一下；
１.Priors类的作用是什么？，另外我看这个类里面self.feature_maps = [127,127,63,63,31,15,15,15]
不应该是self.feature_maps = [12８,12８,6４,6４,3２,1６,1６,1６]吗？
2.训练时数据的增广：我看您的训练实现中训练图像处理为transforms.Resize(512),
        transforms.ToTensor(),
        transforms.Normalize([0.485, 0.456, 0.406], [0.229, 0.224, 0.225])，论文中提到的图像颜色加噪声，反转，随即裁剪是不是没有用到啊？

感谢不吝赐教，谢谢！

**Comment by aoru45** [comment]
@dongfangduoshou123 
抱歉上传的代码并没有完全复现，因为我们现在的任务是球体检测，所以我现在最新的代码是回归球的半径，augmentation的代码也还没有上传，已经上传的代码只是最初版本。上传的代码中anchor匹配也没有按照论文中的做，后面完全写完了会重新上传。

**Comment by dongfangduoshou123** [comment]
响应真快！
没有按照论文的做，自己设计的ａｎｃｈｏｒ－ｆｒｅｅ啊厉害！
重新上传的时候记得也传一份回归框的哦；
后面的实现anchor匹配会按照论文中的做吗？

**Comment by aoru45** [comment]
@dongfangduoshou123 我已经把anchor匹配和论文中loss修改为一样的了。

**Comment by dongfangduoshou123** [comment]
＠aoru45　谢谢，赶紧ｆｏｒｋ一个．

**Comment by dongfangduoshou123** [comment]
提个建议，别人找ＬＦＦＤ的ｐｙｔｏｒｃｈ实现，你这个ｒｅｐｏ名字挺难找到的，改为LFFD-Pytorch之类的，ｓｔａｒ上的能快一些．

**Comment by aoru45** [comment]
@dongfangduoshou123  谢谢你，改了～

**Comment by dongfangduoshou123** [comment]
`def assign_priors(gt_boxes, gt_labels, center_form_priors, iou_threshold=0.6):

    s1 = in_center(center_form_priors.unsqueeze(1),gt_boxes.unsqueeze(0)) # 直接匹配在中间的
    s1[torch.sum(s1, dim=1) > 1] = 0 # 同时有多个匹配的
    best_target_per_prior, best_target_per_prior_index = s1.max(1)
    labels = gt_labels[best_target_per_prior_index]
    labels[best_target_per_prior == False] = 0  # the backgournd id
    boxes = gt_boxes[best_target_per_prior_index]
    t = 0 
    center_form_gt_boxes = corner_form_to_center_form(boxes)
    for f,scale in zip([127,127,63,63,31,15,15,15],[(10,15),(15,20),(20,40),(40,70),(70,110),(110,250),(250,400),(400,560)]):
        # s2 = (center_form_gt_boxes[t:t+f*f,2] > scale[0]/512. * 0.9) & \
        # (center_form_gt_boxes[t:t+f*f,2] < scale[0]/512.) & \
        # (center_form_gt_boxes[t:t+f*f,2] > scale[1]/512.) & \
        # (center_form_gt_boxes[t:t+f*f,2] < scale[1]/512. * 1.1)
        s2 = (center_form_gt_boxes[t:t+f*f, 2] < scale[0]/512.) | \
             (center_form_gt_boxes[t:t+f*f, 2] > scale[1]/512.) 

        s3 = (center_form_gt_boxes[t:t+f*f, 3] < scale[0]/512.) | \
             (center_form_gt_boxes[t:t+f*f, 3] > scale[1]/512.)
        
        labels[t:t+f*f][s2==True] = 0 # 在gray scale内
        labels[t:t+f*f][s3==True] = 0
        t += f*f
    return boxes, labels`

assign_priors函数关于gray scale范围的处理：按照论文中说的比如某个ｂｒａｎｃｈ负责２０－－４０尺寸的人脸，那么１８－２０及４０－－４４尺寸的这个ｂｒａｎｃｈ不管，那么比１８更小或者比４４更大的这个ｂｒａｎｃｈ是不是就更不管了呢．．．另外我看你的这个过滤操作（e.g: center_form_gt_boxes[t:t+f*f, 2]）好像是对ｗ作了过滤center_form_gt_boxes[t:t+f*f, ３］也就是ｈ没有做过滤，是不是ｈｗ都得过滤一下啊？大佬．

**Comment by aoru45** [comment]
您说的有道理，我稍后去修改，咱们共同学习，谢谢～

**Comment by dongfangduoshou123** [comment]
是不是那个LFFDLoss有问题了，我用原作者给的pkl人脸检测数据集训练时，loss到2.左右就不降了;
你有没有发现问题讨论一下啊;他在mxnet里loss不像pytorch里需要将多维tensor reduce成一维再再反传，直接loss的tensor比如是2×159×159的直接就可以backward，这可能是其中一个区别;可以加个微信吗？

**Comment by aoru45** [comment]
@dongfangduoshou123 我这两天有一些学业任务比较麻烦，我会尽快看的，您放心。还有就是，之前我把backbone的普通卷积改成可分离卷积之后收敛非常快。您给个邮箱，我把微信发到您邮箱里，我们以后常交流

**Comment by aoru45** [comment]
loss应该都是一样的会变成一维反传，看这个链接https://discuss.gluon.ai/t/topic/6567

**Comment by dongfangduoshou123** [comment]
好的，争取能原样复现在pytorch中;

我的邮箱：，471747996@qq.com

**Comment by dongfangduoshou123** [comment]
> loss应该都是一样的会变成一维反传，看这个链接https://discuss.gluon.ai/t/topic/6567

这个是gluon里是这样， 他那个用的是mx的symbol，

mask_bbox = mxnet.symbol.slice_axis(mask, axis=1, begin=2, end=6)
predict_bbox = branch_conv3_bbox * mask_bbox
label_bbox = mxnet.symbol.slice_axis(label, axis=1, begin=2, end=6) * mask_bbox
loss_bbox = mxnet.symbol.LinearRegressionOutput(data=predict_bbox, label=label_bbox, name=prefix_name + '_loss_bbox')

return loss_score, loss_bbox

他的上面这个loss_score, loss_bbox 都不是一维的，后面的backward之前，也没有reduce成一维的过程。

**Comment by Dreamgang** [comment]
> @dongfangduoshou123 我这两天有一些学业任务比较麻烦，我会尽快看的，您放心。还有就是，之前我把backbone的普通卷积改成可分离卷积之后收敛非常快。您给个邮箱，我把微信发到您邮箱里，我们以后常交流

你好，请问你是把resblock直接换成深度可分离卷积，还是说只把block里面的普通卷积替换成可分离卷积，谢谢！

**Comment by linwaydong** [comment]
@dongfangduoshou123 @aoru45 能给个联系方式吗？我用keras复现，loss一直降不了，求助下两位大神！！

**Comment by mifan0208** [comment]
@aoru45   你好，我在复现的过程中，由于每个分支都要回传loss,但是loss有很多为0，造成无法回传，并出现cls_loss为Nan,location_loss从很小的2点多开始降正常嘛？，如果方便加了联系方式，谢啦

**Comment by qifenshu** [comment]
> 是不是那个LFFDLoss有问题了，我用原作者给的pkl人脸检测数据集训练时，loss到2.左右就不降了;
> 你有没有发现问题讨论一下啊;他在mxnet里loss不像pytorch里需要将多维tensor reduce成一维再再反传，直接loss的tensor比如是2×159×159的直接就可以backward，这可能是其中一个区别;可以加个微信吗？

你好，请问原作者的pkl人脸检测数据集在哪里可以下载呢？谢谢你。 @dongfangduoshou123
